### ISSUE TITLE
preserveCompilationContext causing failed references for runtime compilation

### ISSUE THREAD

**Issue by Antaris** [dotnet/cli#4047]
Hi,

I am having some issues performing an upgrade from RC2 and RTM. I have a set of libraries deployed to an internal NuGet feed, they have `preserveCompilationContext` set to true. This is mostly because we don't have any clear indication as to whether we are supposed to include that in all projects, or just host projects? `preserveCompilationContext` generates a `.deps.json` file describing the dependency tree for that library.

Now, in my host MVC application, the `MetadataReferenceFeatureProvider` is generating metadata references for the Roslyn-based Razor compiler. 

So, given an example dependency `My.Application.Abstractions`, what I am finding, is that for those assemblies that exist in packages, it is trying to load the `DependencyContext` for each one. Because of the `preserveCompilationContext` setting in the `project.json` project file of the source dependency, there is an entry in the `.deps.json` file:

```
{
  // .deps.json shorted for brevity
  "libraries": {
    "My.Application.Abstractions/1.0.0-master": {
      "type": "project",
      "serviceable": false,
      "sha512": ""
    }
  }
}
```

Now, when iterating through the `DependencyContext.CompileLibraries` property to resolve the reference paths, because the `type` key is set to `project`, it means the `PackageCacheCompilationAssemblyResolver` is skipped (https://github.com/dotnet/cli/blob/rel/1.0.0/src/Microsoft.Extensions.DependencyModel/Resolution/PackageCacheCompilationAssemblyResolver.cs#L39) and in the `AppBaseCompilationAssemblyResolver` (https://github.com/dotnet/cli/blob/rel/1.0.0/src/Microsoft.Extensions.DependencyModel/Resolution/AppBaseCompilationAssemblyResolver.cs#L93) the only directory specified is the application output `./bin/debug/netcoreapp1.0` which doesn't contain the package assembly. Additionally the `PackageCompilationAssemblyResolver` returns `false` because the compilation library is of type `project`. This has a the net result of failing to obtain a reference at all, giving this lovely exception:

`Can not find compilation library location for package`My.Application.Abstractions'

So I have a couple of questions:
1. When using `preserveCompilationContext` in a library, should it be generating `project` as the `type` value?
2. Should we be using `preserveCompilationContext` at all, what are the exact scenarios this _should_ be used and for what types of projects?
3. How do I solve my current issue.

It's difficult to try and generate a working example that I can upload, but if I can, I'll update this issue.

I've cross-posted this over at aspnet/MVC (https://github.com/aspnet/Mvc/issues/5141) as I am unsure as to where this belongs.

**Environment**

```
.NET Command Line Tools (1.0.0-preview2-003121)

Product Information:
 Version:            1.0.0-preview2-003121
 Commit SHA-1 hash:  1e9d529bc5

Runtime Environment:
 OS Name:     Windows
 OS Version:  10.0.14393
 OS Platform: Windows
 RID:         win10-x64
```

**Comment by Antaris** [comment]
This issue was resolved through aspnet/MVC#5141 by removing `preserveCompilationContext`. The scenario was caused through applying `preserveCompilationContext` to a library, pulled in as a package (not a local compilation), and no resolver could be used to locate the assembly.

I'm not sure the likelihood of this becoming  a common problem, but I'm wondering if the `"type": "project` value for the current compilation should be driven by the `emitEntryPoint` flag in `buildOptions`?

**Comment by piotrpMSFT** [comment]
I'm glad the issue was resolved. If there are lingering problems here we'll need to tackle them as they become concrete.
